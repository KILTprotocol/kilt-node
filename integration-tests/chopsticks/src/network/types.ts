import type { SetupOption, setupContext } from '@acala-network/chopsticks-testing'

export type Config = Awaited<ReturnType<typeof setupContext>>

export interface Chain {
	config: (blockNumber?: number, wasmOverride?: string) => SetupOption
	blockNumber?: number
	wasmOverride?: string
	name: string
}

export type ChainConfigs = Record<string, Chain>
