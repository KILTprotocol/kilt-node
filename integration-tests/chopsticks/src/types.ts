export interface AssetSwitchSupplyParameters {
	circulatingSupply: bigint
	sovereignSupply: bigint
	totalSupply: bigint
}
