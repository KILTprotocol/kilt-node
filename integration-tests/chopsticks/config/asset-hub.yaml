endpoint:
  - wss://asset-hub-rococo-rpc.dwellir.com
mock-signature-host: true
runtime-log-level: 5
port: 50001
db: db/asset-hub.sqlite

import-storage:
  System:
    Account:
      -
        -
          - 5Go9qhpqq2EzAZ1SKkkx6bTjnSoTnHEB8FHCnieNS57TimEF
        - providers: 1
          data:
            free: '100000000000000000000000'
