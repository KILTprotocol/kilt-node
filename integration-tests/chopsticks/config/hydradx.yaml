endpoint:
 - wss://hydradx-rpc.dwellir.com
 - wss://rpc.hydradx.cloud
db: ./db/hydradx.db.sqlite
port: 50002
runtime-log-level: 5

import-storage:
 TechnicalCommittee:
  Members:
   - 4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr
 Council:
  Members:
   - 4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr
 System:
  Account:
   -
    -
     - 4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr
    - providers: 1
      data:
       free: "1000000000000000000"
   -
    -
     - 4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ
    - providers: 1
      data:
       free: "1000000000000000000"
 AssetRegistry:
  Assets:
   -
    -
     - 28
    - name: KILT
      assetType: Token
      existentialDeposit: 21358393848783
      symbol: KILT
      decimals: 15
      location:
       parents: 1
       interior: X1(Parachain(2086))
      isSufficient: true
  AssetIds:
   -
    -
     - KILT
    - 28
  AssetLocations:
  -
   -
    - 28
   - parents: 1
     interior:
      X1:
       Parachain: 2086
  LocationAssets:
   -
    -
     - parents: 1
       interior:
        X1:
         Parachain: 2086
    - 28
 MultiTransactionPayment:
  AcceptedCurrencies:
   -
    -
     - 28
    - "72639897479709500000"

