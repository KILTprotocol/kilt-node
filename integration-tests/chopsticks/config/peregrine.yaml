endpoint:
  - wss://peregrine.kilt.io
mock-signature-host: true
db: db/peregrine.sqlite
runtime-log-level: 5
wasm-override: /home/antonio/Developer/kilt-node/target/debug/wbuild/peregrine-runtime/peregrine_runtime.compact.compressed.wasm
