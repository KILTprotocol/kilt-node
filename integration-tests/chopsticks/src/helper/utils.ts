import { ApiPromise, SubmittableResult } from '@polkadot/api'
import { Keyring } from '@polkadot/keyring'
import { u8aToHex } from '@polkadot/util'
import { decodeAddress } from '@polkadot/util-crypto'

import type { SubmittableExtrinsic } from '@polkadot/api/types'
import type { KeyringPair } from '@polkadot/keyring/types'
import type { Codec } from '@polkadot/types/types'
import type { ExpectStatic } from 'vitest'
import { EventRecord } from '@polkadot/types/interfaces'

/**
 * Keyring instance to generate key pairs with 38 ss58Format encoding.
 */
const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

/**
 * Key pairs for Alice, Bob, and Charlie used for testing.
 */
export const keysAlice = keyring.addFromUri('//alice', undefined)
export const keysBob = keyring.addFromUri('//bob', undefined)
export const keysCharlie = keyring.addFromUri('//charlie', undefined)

/**
 * Converts a given value to a Number
 */
export function toNumber(value: string | undefined): number | undefined {
	if (value === undefined) {
		return undefined
	}
	return Number(value)
}

/**
 * Converts a given address to its hexadecimal representation.
 *
 * @param addr - The address to be converted. It should be a string representation of an address.
 */
export function hexAddress(addr: string) {
	return u8aToHex(decodeAddress(addr))
}

/**
 * Validates if the received balance is within the expected range considering a certain precision.
 *
 * @param previousBalance - The balance before the operation.
 * @param receivedBalance - The balance after the operation.
 * @param deltaBalance - The expected change in balance.
 * @param expect - The assertion function from the testing library.
 * @param precision - The precision of the balance validation. It must be a value between 0 and 100.
 *
 * The function calculates the expected balance by adding the deltaBalance to the previousBalance.
 * It then calculates the lower and upper bounds of the expected balance considering the precision.
 * The received balance is then checked if it falls within the range of the lower and upper bounds.
 *
 * If the precision is not between 0 and 100, the function throws an error.
 *
 * @throws {Error} If the precision is not between 0 and 100.
 */
export function validateBalanceWithPrecision(
	previousBalance: bigint,
	receivedBalance: bigint,
	deltaBalance: bigint,
	expect: ExpectStatic,
	precision: bigint
) {
	if (precision < 0n || precision > 100n) {
		throw new Error('Precision must be between 0 and 100')
	}

	const allowedError = 100n - precision
	const expectedBalance = previousBalance + deltaBalance

	const lowerBound = expectedBalance - (expectedBalance * allowedError) / 100n
	const upperBound = expectedBalance + (expectedBalance * allowedError) / 100n

	expect(receivedBalance).toBeGreaterThanOrEqual(lowerBound)
	expect(receivedBalance).toBeLessThanOrEqual(upperBound)
}

/**
 * Fetches the paid fees for the executed XCM message. Is only working on the sender chain.
 */
export async function getPaidXcmFees(api: ApiPromise, eventRecords: Codec[]): Promise<bigint> {
	const polkadotFees = eventRecords.find((eventRecord) =>
		api.events.polkadotXcm.FeesPaid.is((eventRecord as EventRecord).event)
	)
	if (!polkadotFees) {
		throw new Error('FeesPaid event not found')
	}

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	return (polkadotFees as any).event.data.fees[0].fun.asFungible.toBigInt()
}

/**
 * Calculates the transaction fees for a given transaction.
 */
export async function calculateTxFees(
	tx: SubmittableExtrinsic<'promise', SubmittableResult>,
	account: KeyringPair
): Promise<bigint> {
	const paymentInfo = await tx.paymentInfo(account)
	return paymentInfo.partialFee.toBigInt()
}

// Constants for the balances
export const KILT = BigInt(1e15)
export const DOT = BigInt(1e10)
export const HDX = BigInt(1e12)

export const initialBalanceKILT = 100n * KILT
export const initialBalanceDOT = 100n * DOT
export const initialBalanceHDX = 100n * HDX
