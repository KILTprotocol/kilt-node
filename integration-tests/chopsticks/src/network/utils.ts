export function getSiblingLocation(paraId: number) {
	return {
		parents: 1,
		interior: {
			X1: { Parachain: paraId },
		},
	}
}

export function getParentLocation() {
	return {
		parents: 1,
		interior: 'Here',
	}
}

export function getAccountLocationV2(addr: string) {
	return {
		V2: {
			parents: 0,
			interior: {
				X1: {
					AccountId32: {
						network: 'Any',
						id: addr,
					},
				},
			},
		},
	}
}

export function getAccountLocationV3(addr: string) {
	return {
		V3: {
			parents: 0,
			interior: {
				X1: {
					AccountId32: {
						id: addr,
					},
				},
			},
		},
	}
}

export function getNativeAssetIdLocation(amount: bigint) {
	return {
		id: { Concrete: { parents: 0, interior: 'Here' } },
		fun: { Fungible: amount },
	}
}
