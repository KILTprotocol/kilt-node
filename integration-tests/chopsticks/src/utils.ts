import { Keyring } from '@polkadot/keyring'

import { AssetSwitchSupplyParameters } from './types.js'

const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

export const keysAlice = keyring.addFromUri('//alice', undefined, 'ed25519')
export const keysBob = keyring.addFromUri('//bob', undefined, 'ed25519')
export const keysCharlie = keyring.addFromUri('//charlie', undefined, 'ed25519')

export function toNumber(value: string | undefined): number | undefined {
	if (value === undefined) {
		return undefined
	}

	return Number(value)
}

export function getAssetSwitchParameters(
	totalSupply: bigint = BigInt('100000000000000000000'),
	ratioCirculating: bigint = BigInt(10)
): AssetSwitchSupplyParameters {
	const circulatingSupply = (totalSupply * ratioCirculating) / BigInt(100)
	const sovereignSupply = totalSupply - circulatingSupply
	return { circulatingSupply, sovereignSupply, totalSupply }
}

export const KILT = BigInt(1e15)
export const DOT = BigInt(1e10)
export const HDX = BigInt(1e12)
export const ROC = BigInt(1e12)

export const initialBalanceKILT = BigInt(100) * KILT
export const initialBalanceDOT = BigInt(100) * DOT
export const initialBalanceHDX = BigInt(100) * HDX
export const initialBalanceROC = BigInt(100) * ROC
