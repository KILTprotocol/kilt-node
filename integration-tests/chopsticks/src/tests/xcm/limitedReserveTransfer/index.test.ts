/* eslint-disable @typescript-eslint/no-explicit-any */
import { describe, beforeEach, it, afterEach, ExpectStatic } from 'vitest'
import { sendTransaction, withExpect } from '@acala-network/chopsticks-testing'

import { createBlock, hexAddress, setStorage } from '../../utils.js'

import { testPairsLimitedReserveTransfers } from './config.js'
import { Config } from '../../../network/types.js'
import { setupNetwork, shutDownNetwork } from '../../index.js'
import type { KeyringPair } from '@polkadot/keyring/types'

describe.each(testPairsLimitedReserveTransfers)(
	'Limited Reserve Transfers',
	{ sequential: true, timeout: 30_000 },
	async ({ blockchain, storage, accounts, query, sovereignAccount, test, config }) => {
		let senderContext: Config
		let receiverContext: Config
		let relayContext: Config
		let senderAccount: KeyringPair
		let receiverAccount: KeyringPair
		const { desc, precision } = config

		beforeEach(async () => {
			const { receiver, sender, relay } = blockchain

			const { receiverChainContext, senderChainContext, relayChainContext } = await setupNetwork(
				relay,
				sender,
				receiver
			)

			relayContext = relayChainContext
			senderContext = senderChainContext
			receiverContext = receiverChainContext

			const { receiverStorage, senderStorage, relayStorage } = storage
			await setStorage(senderContext, senderStorage)
			await setStorage(receiverContext, receiverStorage)
			await setStorage(relayContext, relayStorage)

			const { senderAccount: a, receiverAccount: b } = accounts
			senderAccount = a
			receiverAccount = b
		}, 20_000)

		afterEach(async () => {
			try {
				await shutDownNetwork([senderContext, receiverContext, relayContext])
			} catch (error) {
				console.error(error)
			}
		})

		it(desc, { timeout: 10_000, retry: 3 }, async ({ expect }) => {
			const { checkEvents, checkSystemEvents } = withExpect(expect)

			// test parameters
			const { pallets, tx, balanceToTransfer } = test

			// Balance of the receiver sovereign account before the transfer
			const receiverSovereignAccountBalanceBeforeTransfer = await query.sender(
				senderContext,
				sovereignAccount.sender
			)

			const initialBalanceReceiver = await query.receiver(receiverContext, receiverAccount.address)

			// Check initial balance receiver should be zero
			expect(initialBalanceReceiver).toBe(BigInt(0))

			const signedTx = tx(
				senderContext,
				hexAddress(receiverAccount.address),
				balanceToTransfer.toString()
			).signAsync(senderAccount)

			const events = await sendTransaction(signedTx)

			// check sender state
			await createBlock(senderContext)

			pallets.sender.map((pallet) =>
				checkEvents(events, pallet).toMatchSnapshot(`sender events ${JSON.stringify(pallet)}`)
			)

			const balanceSenderAfterTransfer = await query.sender(senderContext, senderAccount.address)
			const receiverSovereignAccountBalanceAfterTransfer = await query.sender(
				senderContext,
				sovereignAccount.sender
			)
			expect(receiverSovereignAccountBalanceAfterTransfer).toBe(
				receiverSovereignAccountBalanceBeforeTransfer + BigInt(balanceToTransfer)
			)

			checkBalanceInRange(balanceSenderAfterTransfer, expect, precision)

			// check receiver state
			await createBlock(receiverContext)

			pallets.receiver.map((pallet) =>
				checkSystemEvents(receiverContext, pallet).toMatchSnapshot(`receiver events ${JSON.stringify(pallet)}`)
			)

			const balanceReceiverAfterTransfer = await query.receiver(receiverContext, receiverAccount.address)

			checkBalanceInRange(balanceReceiverAfterTransfer, expect, precision)
		})
	}
)

// Check Balance is in range
function checkBalanceInRange(receivedBalance: bigint, expect: ExpectStatic, precision: bigint) {
	if (precision < BigInt(0) || precision > BigInt(100)) {
		throw new Error('Precision must be between 0 and 100')
	}

	const lowerBound = (receivedBalance * precision) / BigInt(100)
	expect(receivedBalance).toBeLessThanOrEqual(receivedBalance)
	expect(receivedBalance).toBeGreaterThanOrEqual(lowerBound)
}
