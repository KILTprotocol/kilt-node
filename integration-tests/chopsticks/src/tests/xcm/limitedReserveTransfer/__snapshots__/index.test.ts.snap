// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Limited Reserve Transfers > Kilt -> HydraDx at block > receiver events "currencies" 1`] = `
[
  {
    "data": {
      "amount": "(rounded 1000000000000000)",
      "currencyId": 28,
      "who": "7MZG43idRmdg8VSt5BS9mVJeBhhxxt5y55hCsMpoKp5xFQX2",
    },
    "method": "Deposited",
    "section": "currencies",
  },
  {
    "data": {
      "amount": "(rounded 4100000000000)",
      "currencyId": 28,
      "who": "7L53bUTBopuwFt3mKUfmkzgGLayYa1Yvn1hAg9v5UMrQzTfh",
    },
    "method": "Deposited",
    "section": "currencies",
  },
]
`;

exports[`Limited Reserve Transfers > Kilt -> HydraDx at block > receiver events "tokens" 1`] = `
[
  {
    "data": {
      "amount": "(rounded 1000000000000000)",
      "currencyId": 28,
      "who": "7MZG43idRmdg8VSt5BS9mVJeBhhxxt5y55hCsMpoKp5xFQX2",
    },
    "method": "Endowed",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 1000000000000000)",
      "currencyId": 28,
      "who": "7MZG43idRmdg8VSt5BS9mVJeBhhxxt5y55hCsMpoKp5xFQX2",
    },
    "method": "Deposited",
    "section": "tokens",
  },
  {
    "data": {
      "amount": "(rounded 4100000000000)",
      "currencyId": 28,
      "who": "7L53bUTBopuwFt3mKUfmkzgGLayYa1Yvn1hAg9v5UMrQzTfh",
    },
    "method": "Deposited",
    "section": "tokens",
  },
]
`;

exports[`Limited Reserve Transfers > Kilt -> HydraDx at block > receiver events "xcmpQueue" 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
      "messageId": "(hash)",
      "weight": {
        "proofSize": 0,
        "refTime": 400000000,
      },
    },
    "method": "Success",
    "section": "xcmpQueue",
  },
]
`;

exports[`Limited Reserve Transfers > Kilt -> HydraDx at block > sender events "polkadotXcm" 1`] = `
[
  {
    "data": {
      "outcome": {
        "Complete": {
          "proofSize": 0,
          "refTime": 400000000,
        },
      },
    },
    "method": "Attempted",
    "section": "polkadotXcm",
  },
]
`;

exports[`Limited Reserve Transfers > Kilt -> HydraDx at block > sender events "xcmpQueue" 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
]
`;

exports[`Limited Reserve Transfers > Kilt -> HydraDx at block > sender events {"section":"balances","method":"Withdraw"} 1`] = `
[
  {
    "data": {
      "amount": "(rounded 170000000000)",
      "who": "4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ",
    },
    "method": "Withdraw",
    "section": "balances",
  },
]
`;

exports[`Limited Reserve Transfers > Kilt -> HydraDx live status > receiver events "xcmpQueue" 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
      "messageId": "(hash)",
      "weight": {
        "proofSize": 0,
        "refTime": 400000000,
      },
    },
    "method": "Success",
    "section": "xcmpQueue",
  },
]
`;

exports[`Limited Reserve Transfers > Kilt -> HydraDx live status > sender events "polkadotXcm" 1`] = `
[
  {
    "data": {
      "outcome": {
        "Complete": {
          "proofSize": 0,
          "refTime": 400000000,
        },
      },
    },
    "method": "Attempted",
    "section": "polkadotXcm",
  },
]
`;

exports[`Limited Reserve Transfers > Kilt -> HydraDx live status > sender events "xcmpQueue" 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
]
`;
